# **CSCE 5218 / CSCE 4930 Deep Learning**

# **HW1a The Perceptron** (20 pt)


In [83]:
# Get the datasets
!curl.exe http://huang.eng.unt.edu/CSCE-5218/train.dat
!curl.exe http://huang.eng.unt.edu/CSCE-5218/test.dat


A1	A2	A3	A4	A5	A6	A7	A8	A9	A10	A11	A12	A13	

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 11244  100 11244    0     0  10305      0  0:00:01  0:00:01 --:--:-- 11034



1	1	0	0	0	0	0	0	1	1	0	0	1	0
0	0	1	1	0	1	1	0	0	0	0	0	1	0
0	1	0	1	1	0	1	0	1	1	1	0	1	1
0	0	1	0	0	1	0	1	0	1	1	1	1	0
0	1	0	0	0	0	0	1	1	1	1	1	1	0
0	1	1	1	0	0	0	1	0	1	1	0	1	1
0	1	1	0	0	0	1	0	0	0	0	0	1	0
0	0	0	1	1	0	1	1	1	0	0	0	1	0
0	0	0	0	0	0	1	0	1	0	1	0	1	0
1	0	1	1	1	0	0	0	1	1	1	0	1	1
0	1	1	0	1	0	1	1	1	1	1	1	1	1
0	0	1	1	0	0	1	0	0	1	1	1	1	0
1	0	1	1	0	0	1	1	0	1	0	0	1	1
0	0	1	1	0	0	0	0	1	0	1	1	1	0
1	0	0	0	0	0	1	0	1	1	0	1	1	0
0	1	1	1	1	1	0	0	1	0	1	0	1	1
1	0	0	0	1	0	1	1	1	1	1	1	1	0
1	0	0	0	0	1	1	0	1	1	1	1	1	0
1	0	0	1	0	1	0	1	1	0	0	1	1	0
0	0	1	1	0	1	1	0	0	0	0	1	1	0
1	1	1	0	0	1	1	1	1	1	0	1	1	1
0	1	1	1	1	1	0	0	0	1	0	1	1	0
0	0	1	1	0	1	0	1	0	1	1	1	1	0
1	0	0	1	0	1	1	1	1	0	0	1	1	0
0	0	1	0	0	0	1	0	0	0	0	0	1	0
1	0	1	0	0	0	1	1	0	0	0	1	1	0
1	0	0	0	0	1	0	0	0	1	0	1	1	0
1	0	0	0	1	1	1	0	0	1	0	1	1	0
1	0	0	0	0	0	1	1	1	1	1	0	1	0
0	0	1	1	0	1	0	1	0	0	1	0	1	0
0	1	1	0	0	0	1	1	1	1	0	1	1	0
0	0	0	0	1	1	1	1	0	0	0	0	1	0
0	1	1	0	1	1	1	1	0	0	0	1	1	0
0	1	0	1	0	0	0	1	0	0	0	1	1	0
1	0	0	1	0	0	1	1	1	1	0	0	1	0
0	0	0	1	0	1	1	1	1	1

A1	A2	A3	A4	A5	A6	A7	A8	A9	A10	A11	A12	A13

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  2844  100  2844    0     0   5677      0 --:--:-- --:--:-- --:--:-- 17132
100  2844  100  2844    0     0   5670      0 --:--:-- --:--:-- --:--:-- 17132



1	1	1	1	0	0	1	1	0	0	0	1	1	0
0	0	0	1	0	0	1	1	0	1	0	0	1	0
0	1	1	1	0	1	1	1	1	0	0	0	1	0
0	1	1	0	1	0	1	1	1	0	1	0	1	0
0	1	0	0	0	1	0	1	0	1	0	0	1	0
0	1	1	0	0	1	1	1	1	1	1	0	1	0
0	1	1	1	0	0	1	1	0	0	0	1	1	0
0	1	0	0	1	0	0	1	1	0	1	1	1	0
1	1	1	1	0	0	1	1	0	0	0	0	1	0
1	0	1	1	1	1	1	0	0	0	0	0	1	0
1	0	0	0	1	0	0	1	1	0	1	1	1	0
0	0	1	0	0	1	1	0	0	0	0	1	1	0
0	1	0	1	1	1	0	0	0	0	0	1	1	0
0	1	0	0	0	1	1	1	0	1	1	1	1	0
1	1	1	0	1	0	0	1	0	1	1	0	1	1
1	0	0	1	0	0	1	1	0	0	1	1	1	0
0	0	0	1	0	1	1	0	1	1	1	1	1	0
1	0	1	0	1	1	0	0	0	1	0	0	1	0
0	1	1	0	1	0	1	0	1	1	1	0	1	1
0	1	0	0	1	1	1	1	1	1	0	0	1	0
1	1	0	1	0	1	1	0	1	0	0	0	1	0
1	0	1	1	0	1	0	0	0	0	1	1	1	1
1	0	1	1	1	1	0	1	1	0	1	0	1	1
1	0	0	1	0	0	0	1	0	1	0	1	1	1
1	1	0	1	1	0	1	1	0	0	1	0	1	0
0	0	0	0	1	1	1	0	0	1	0	1	1	0
1	1	1	0	0	0	1	0	0	0	1	1	1	0
1	0	1	0	0	1	0	0	0	0	1	1	1	0
1	0	1	1	1	1	0	0	0	0	1	1	1	0
0	1	0	0	1	1	1	0	1	1	0	1	1	1
0	1	1	1	0	0	1	0	1	1	1	0	1	1
1	0	0	0	0	0	0	1	1	1	1	0	1	1
1	0	1	1	0	1	1	0	0	1	0	1	1	0
1	1	1	0	1	1	0	1	1	0	1	0	1	1
0	1	0	1	0	1	0	1	0	0	0	1	1	0
1	0	0	1	0	1	0	1	1	1

In [84]:
# Take a peek at the datasets
!curl --output train.dat http://huang.eng.unt.edu/CSCE-5218/train.dat
!curl --output test.dat http://huang.eng.unt.edu/CSCE-5218/test.dat

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 11244  100 11244    0     0  39568      0 --:--:-- --:--:-- --:--:-- 41338
100 11244  100 11244    0     0  39517      0 --:--:-- --:--:-- --:--:-- 41186
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  2844  100  2844    0     0   3835      0 --:--:-- --:--:-- --:--:-- 11421


### Build the Perceptron Model

You will need to complete some of the function definitions below.  DO NOT import any other libraries to complete this. 

In [87]:
import math
import itertools
import re

# Corpus reader, all columns but the last one are coordinates;
#   the last column is the label
def read_data(file_name):
    f = open(file_name, 'r')

    data = []
    # Discard header line
    f.readline()
    for instance in f.readlines():
        if not re.search('\t', instance): continue
        instance = list(map(int, instance.strip().split('\t')))
        # Add a dummy input so that w0 becomes the bias
        #print(instance)
        #print("after readline")
        instance = [-1] + instance
        #print(instance)
        data += [instance]
    return data


def dot_product(array1, array2):
    dot_product = 0
    for i,j in zip(array1,array2):
        dot_product = dot_product + (i * j)
    return dot_product


def sigmoid(x):
    return (1.0 / (1.0 + math.exp(-1.0*x)))

# The output of the model, which for the perceptron is 
# the sigmoid function applied to the dot product of 
# the instance and the weights
def output(weight, instance):
    output = sigmoid(dot_product(weight, instance))
    return output

# Predict the label of an instance; this is the definition of the perceptron
# you should output 1 if the output is >= 0.5 else output 0
def predict(weights, instance):
    if output(weights, instance) >= 0.5:
        return 1
    else:
        return 0

# Accuracy = percent of correct predictions
def get_accuracy(weights, instances):
    # You do not to write code like this, but get used to it
    correct = sum([1 if predict(weights, instance) == instance[-1] else 0
                   for instance in instances])
    return correct * 100 / len(instances)


# Train a perceptron with instances and hyperparameters:
#       lr (learning rate) 
#       epochs
# The implementation comes from the definition of the perceptron
#
# Training consists on fitting the parameters which are the weights
# that's the only thing training is responsible to fit
# (recall that w0 is the bias, and w1..wn are the weights for each coordinate)
#
# Hyperparameters (lr and epochs) are given to the training algorithm
# We are updating weights in the opposite direction of the gradient of the error,
# so with a "decent" lr we are guaranteed to reduce the error after each iteration.
def train_perceptron(instances, lr, epochs):

    #TODO: name this step
    weights = [0] * (len(instances[0]))
    #print(weights)
    for _ in range(epochs):
        for instance in instances:
            #TODO: name these steps
            #print(instance)
            #print(weights)
            in_value = dot_product(weights, instance)
            #print(in_value)
            #print("--------------------")
            output = sigmoid(in_value)
            error = instance[-1] - output
            #TODO: name these steps
            for i in range(0, len(weights)):
                weights[i] += lr * error * output * (1-output) * instance[i]

    return weights

## Run it

In [88]:
instances_tr = read_data("train.dat")
instances_te = read_data("test.dat")
lr = 0.005
epochs = 5

#print(instances_tr[0])

weights = train_perceptron(instances_tr, lr, epochs)
print(weights)
accuracy = get_accuracy(weights, instances_te)
print(f"#tr: {len(instances_tr):3}, epochs: {epochs:3}, learning rate: {lr:.3f}; "
      f"Accuracy (test, {len(instances_te)} instances): {accuracy:.1f}")

[0.19834690842618988, -0.047528630608296846, -0.07445715046174368, -0.05847954338095892, -0.06206714117763581, -0.11731257435626952, -0.10312217725648434, -0.1786193849564548, -0.06766220948749965, -0.01690275153925823, -0.025589272102902735, 0.05420153371311189, -0.14840632708386856, -0.19834690842618988, 0.3941623144280716]
#tr: 400, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0


## Questions

Answer the following questions. Include your implementation and the output for each question.



### Question 1

In `train_perceptron(instances, lr, epochs)`, we have the follosing code:
```
in_value = dot_product(weights, instance)
output = sigmoid(in_value)
error = instance[-1] - output
```

Why don't we have the following code snippet instead?
```
output = predict(weights, instance)
error = instance[-1] - output
```

#### Answer
At each instance, the input is calculated based on the weight associated with it. Then the sigmoid function needs to be applied on the input to calculate the gradient decent to finally calculate the error. This process repeats until the error is at the required minimum threshold.

So using the the following code snippet,
in_value = dot_product(weights, instance)
output = sigmoid(in_value)
error = instance[-1] - output

we are calculating more accurate weights which yeilds better learning rate and accuracy.

weights:
[0.19834690842618988, -0.047528630608296846, -0.07445715046174368, -0.05847954338095892, -0.06206714117763581, -0.11731257435626952, -0.10312217725648434, -0.1786193849564548, -0.06766220948749965, -0.01690275153925823, -0.025589272102902735, 0.05420153371311189, -0.14840632708386856, -0.19834690842618988, 0.3941623144280716]
output:
#tr: 400, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0

When using the following code snippet, the weights are not calculated correctly.
output = predict(weights, instance)
error = instance[-1] - output

weights:
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]


### Question 2
Train the perceptron with the following hyperparameters and calculate the accuracy with the test dataset.

```
tr_percent = [5, 10, 25, 50, 75, 100] # percent of the training dataset to train with
num_epochs = [5, 10, 20, 50, 100]              # number of epochs
lr = [0.005, 0.01, 0.05]              # learning rate
```

TODO: Write your code below and include the output at the end of each training loop (NOT AFTER EACH EPOCH)
of your code.The output should look like the following:
```
# tr:  20, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
# tr:  20, epochs:  10, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
# tr:  20, epochs:  20, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
[and so on for all the combinations]
```
You will get different results with different hyperparameters.

#### TODO Add your answer here (code and output in the format above) 
Hyperparameters:
tr_percent = [5, 10, 25, 50, 75, 100] # percent of the training dataset to train with
num_epochs = [5, 10, 20, 50, 100]     # number of epochs
lr_array = [0.005, 0.01, 0.05]        # learning rate

Output:
#tr: 20, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 74.0
#tr: 40, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
#tr: 100, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 92.0
#tr: 200, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 100.0
#tr: 300, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 100.0
#tr: 400, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 100.0
#tr: 20, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 85.0
#tr: 40, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 80.0
#tr: 100, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 98.0
#tr: 200, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 100.0
#tr: 300, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 100.0
#tr: 400, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 100.0
#tr: 20, epochs: 100, learning rate: 0.050; Accuracy (test, 100 instances): 91.0
#tr: 40, epochs: 100, learning rate: 0.050; Accuracy (test, 100 instances): 97.0
#tr: 100, epochs: 100, learning rate: 0.050; Accuracy (test, 100 instances): 100.0
#tr: 200, epochs: 100, learning rate: 0.050; Accuracy (test, 100 instances): 100.0
#tr: 300, epochs: 100, learning rate: 0.050; Accuracy (test, 100 instances): 100.0
#tr: 400, epochs: 100, learning rate: 0.050; Accuracy (test, 100 instances): 100.0


In [89]:
instances_tr = read_data("train.dat")
instances_te = read_data("test.dat")
tr_percent = [5, 10, 25, 50, 75, 100] # percent of the training dataset to train with
num_epochs = [5, 10, 20, 50, 100]     # number of epochs
lr_array = [0.005, 0.01, 0.05]        # learning rate

for lr in lr_array:
  for tr_size in tr_percent:
    for epochs in num_epochs:
      size =  round(len(instances_tr)*tr_size/100)
      pre_instances = instances_tr[0:size]
      weights = train_perceptron(pre_instances, lr, epochs)
      accuracy = get_accuracy(weights, instances_te)
    print(f"#tr: {len(pre_instances):0}, epochs: {epochs:3}, learning rate: {lr:.3f}; "
            f"Accuracy (test, {len(instances_te)} instances): {accuracy:.1f}")

#tr: 20, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 74.0
#tr: 40, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
#tr: 100, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 92.0
#tr: 200, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 100.0
#tr: 300, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 100.0
#tr: 400, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 100.0
#tr: 20, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 85.0
#tr: 40, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 80.0
#tr: 100, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 98.0
#tr: 200, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 100.0
#tr: 300, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 100.0
#tr: 400, epochs: 100, learning rate: 0.010; Accuracy (test, 100 instances): 100.0
#tr: 20, epoch

### Question 3
Write a couple paragraphs interpreting the results with all the combinations of hyperparameters. Drawing a plot will probably help you make a point. In particular, answer the following:
- A. Do you need to train with all the training dataset to get the highest accuracy with the test dataset?
     Answer: The best practise is to use the full training dataset to train the model so the model can learn complete charecteristcs of the data. Sometimes, the full dataset can be massive and can take large amount of time. So, one can make the judgement on how much data is a good representation of the training data inorder to achieve desired accuracy. The recommended approach is to allocate 80% of training dataset for traning and 20% for testing.
     
- B. How do you justify that training the second run obtains worse accuracy than the first one (despite the second one uses more training data)?
     Answer: Clearly, the second run used twice the size of training data compared to first run but the second run yielded lower accuracy than first. This indicates that more training data doesn't always mean better accuracy. It implies that the model have been overfit and not performing well to new data.
     The hyperparameters used also makes a difference on the accuracy. The lr parameter value may also played a role in lower accuracy of the second run. So one should fine tune the hyperparameters to attain reasonable accuracy.
   ```
#tr: 100, epochs:  20, learning rate: 0.050; Accuracy (test, 100 instances): 71.0
#tr: 200, epochs:  20, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
```
- C. Can you get higher accuracy with additional hyperparameters (higher than `80.0`)?
     Answer: Additional hyperparameters may not necessarily get higher accuracy. By fine tuning the hyperparameters we may be able to achieve higher accuracy.
     
- D. Is it always worth training for more epochs (while keeping all other hyperparameters fixed)?
     Answer: It is not always a good idea to train the model for more epochs while keeping all other hyperparameters. Fine tuning the model performance is an iterative process. This is achieved by trying out various combinations of hyperparameters to achieve desired accuracy. 
     
#### TODO: Add your answer here (code and text)

